File: src/ingestion/chains/ktc.ts
/**
 * KTC Chain Adapter
 *
 * Adapter for parsing KTC retail chain price data files.
 * KTC uses CSV format with semicolon delimiter and Windows-1250 encoding.
 * Store resolution is based on filename.
 *
 * KTC portal: https://www.ktc.hr/cjenici
 *
 * Portal structure:
 * 1. Main page lists all stores/regions (poslovnica)
 * 2. Each store page lists CSV files by date
 * 3. CSV files are at /ktcftp/Cjenici/STORE_NAME/FILENAME.csv
 *
 * Filename format: TRGOVINA-ADDRESS-STORE_ID-DATE-TIME.csv
 * Example: TRGOVINA-PAKRACKA ULICA 1 BJELOVAR-PJ50-1-20260105-071001.csv
 */

import type { DiscoveredFile, StoreMetadata } from '../core/types'
import type { CsvColumnMapping } from '../parsers/csv'
import { BaseCsvAdapter } from './base'
import { CHAIN_CONFIGS } from './config'

/**
 * Column mapping for KTC CSV files.
 * Maps KTC's column names to NormalizedRow fields.
 *
 * Actual KTC CSV headers (Windows-1250 encoded):
 * - Naziv proizvoda
 * - Šifra proizvoda
 * - Marka proizvoda
 * - Neto količina
 * - Jedinica mjere
 * - Maloprodajna cijena
 * - Cijena za jedinicu mjere
 * - Barkod
 * - Kategorija
 * - Najniža cijena u posljednjih 30 dana
 * - MPC za vrijeme posebnog oblika prodaje
 */
const KTC_COLUMN_MAPPING: CsvColumnMapping = {
  name: 'Naziv proizvoda',
  externalId: 'Šifra proizvoda',
  brand: 'Marka proizvoda',
  unitQuantity: 'Neto količina',
  unit: 'Jedinica mjere',
  price: 'Maloprodajna cijena',
  unitPrice: 'Cijena za jedinicu mjere',
  barcodes: 'Barkod',
  category: 'Kategorija',
  lowestPrice30d: 'Najniža cijena u posljednjih 30 dana',
  discountPrice: 'MPC za vrijeme posebnog oblika prodaje',
}

/**
 * Alternative column mapping for KTC CSV files (without diacritics).
 * Some systems may strip diacritics when decoding Windows-1250.
 */
const KTC_COLUMN_MAPPING_ALT: CsvColumnMapping = {
  name: 'Naziv proizvoda',
  externalId: 'Sifra proizvoda',
  brand: 'Marka proizvoda',
  unitQuantity: 'Neto kolicina',
  unit: 'Jedinica mjere',
  price: 'Maloprodajna cijena',
  unitPrice: 'Cijena za jedinicu mjere',
  barcodes: 'Barkod',
  category: 'Kategorija',
  lowestPrice30d: 'Najniza cijena u posljednjih 30 dana',
  discountPrice: 'MPC za vrijeme posebnog oblika prodaje',
}

/**
 * KTC chain adapter implementation.
 * Extends BaseCsvAdapter for common CSV parsing functionality.
 *
 * KTC uses a two-level portal structure:
 * 1. Main page lists all stores (poslovnica)
 * 2. Each store page lists downloadable CSV files by date
 */
export class KtcAdapter extends BaseCsvAdapter {
  /** Date to discover files for (YYYY-MM-DD format, set by CLI before discovery) */
  private discoveryDate: string | null = null

  constructor() {
    super({
      slug: 'ktc',
      name: 'KTC',
      supportedTypes: ['csv'],
      chainConfig: CHAIN_CONFIGS.ktc,
      columnMapping: KTC_COLUMN_MAPPING,
      alternativeColumnMapping: KTC_COLUMN_MAPPING_ALT,
      filenamePrefixPatterns: [
        /^TRGOVINA[_-]?/i,
        /^KTC[_-]?/i,
        /^cjenik[_-]?/i,
      ],
      rateLimitConfig: {
        requestsPerSecond: 2,
        maxRetries: 3,
      },
    })
  }

  /**
   * Set the date to use for discovery.
   * @param date - Date in YYYY-MM-DD format
   */
  setDiscoveryDate(date: string): void {
    this.discoveryDate = date
  }

  /**
   * Extract date from KTC filename.
   * KTC filenames contain dates in format: YYYYMMDD
   * Example: TRGOVINA-PAKRACKA ULICA 1 BJELOVAR-PJ50-1-20260105-071001.csv
   */
  private extractDateFromFilename(filename: string): string | null {
    // Match 8-digit date pattern (YYYYMMDD)
    const match = filename.match(/(\d{4})(\d{2})(\d{2})-\d{6}\.csv$/i)
    if (match) {
      return `${match[1]}-${match[2]}-${match[3]}`
    }
    return null
  }

  /**
   * Discover available KTC price files.
   *
   * KTC's portal at /cjenici has a two-level structure:
   * 1. Main page lists stores (e.g., ?poslovnica=RC%20BJELOVAR%20PJ-50)
   * 2. Store pages list CSV files with download links
   *
   * This method:
   * 1. Fetches the main page to get all store names
   * 2. For each store, fetches the store page to get CSV links
   * 3. Filters by date if setDiscoveryDate was called
   *
   * @returns Array of discovered files (filtered by date if set)
   */
  async discover(): Promise<DiscoveredFile[]> {
    const discoveredFiles: DiscoveredFile[] = []
    const seenUrls = new Set<string>()

    console.log(`[DEBUG] Fetching ${this.name} portal: ${this.config.baseUrl}`)

    try {
      // Step 1: Fetch main page to get list of stores
      const mainResponse = await fetch(this.config.baseUrl, {
        headers: {
          'User-Agent': 'Mozilla/5.0 (compatible; PriceTracker/1.0)',
          'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
        },
      })

      if (!mainResponse.ok) {
        console.error(`[ERROR] Failed to fetch KTC portal: ${mainResponse.status} ${mainResponse.statusText}`)
        return []
      }

      const mainHtml = await mainResponse.text()

      // Extract store names from links like: ?poslovnica=RC BJELOVAR PJ-50
      const storePattern = /poslovnica=([^"&]+)/g
      const stores: string[] = []
      let storeMatch: RegExpExecArray | null

      while ((storeMatch = storePattern.exec(mainHtml)) !== null) {
        const storeName = decodeURIComponent(storeMatch[1])
        if (!stores.includes(storeName)) {
          stores.push(storeName)
        }
      }

      console.log(`[DEBUG] Found ${stores.length} store(s) on KTC portal`)

      // Step 2: For each store, fetch store page and extract CSV links
      for (const storeName of stores) {
        const storeUrl = `${this.config.baseUrl}?poslovnica=${encodeURIComponent(storeName)}`

        try {
          // Rate limit between store requests
          await this.rateLimiter.throttle()

          const storeResponse = await fetch(storeUrl, {
            headers: {
              'User-Agent': 'Mozilla/5.0 (compatible; PriceTracker/1.0)',
              'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
            },
          })

          if (!storeResponse.ok) {
            console.warn(`[WARN] Failed to fetch store page for ${storeName}: ${storeResponse.status}`)
            continue
          }

          const storeHtml = await storeResponse.text()

          // Extract CSV links like: /ktcftp/Cjenici/STORE_NAME/FILENAME.csv
          const csvPattern = /href="([^"]*\.csv)"/gi
          let csvMatch: RegExpExecArray | null

          while ((csvMatch = csvPattern.exec(storeHtml)) !== null) {
            const href = csvMatch[1]
            const fileUrl = href.startsWith('http') ? href : new URL(href, this.config.baseUrl).toString()

            // Skip duplicates
            if (seenUrls.has(fileUrl)) {
              continue
            }
            seenUrls.add(fileUrl)

            const filename = this.extractFilenameFromUrl(fileUrl)
            const fileDate = this.extractDateFromFilename(filename)

            // Filter by date if discoveryDate is set
            if (this.discoveryDate && fileDate && fileDate !== this.discoveryDate) {
              continue
            }

            discoveredFiles.push({
              url: fileUrl,
              filename,
              type: 'csv',
              size: null,
              lastModified: fileDate ? new Date(fileDate) : null,
              metadata: {
                source: 'ktc_portal',
                discoveredAt: new Date().toISOString(),
                storeName,
                ...(fileDate && { portalDate: fileDate }),
              },
            })
          }
        } catch (error) {
          const errorMessage = error instanceof Error ? error.message : String(error)
          console.warn(`[WARN] Error fetching store ${storeName}: ${errorMessage}`)
        }
      }

      return discoveredFiles
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error)
      console.error(`[ERROR] Error discovering KTC files: ${errorMessage}`)
      return []
    }
  }

  /**
   * Extract store identifier from KTC filename.
   *
   * KTC filenames follow the pattern:
   * TRGOVINA-ADDRESS-STORE_ID-DATE-TIME.csv
   *
   * Example: TRGOVINA-PAKRACKA ULICA 1 BJELOVAR-PJ50-1-20260105-071001.csv
   * Store ID would be: PJ50-1
   *
   * Store IDs can include letters after PJ: PJ06, PJ7B, PJ8A, etc.
   *
   * @param filename - The CSV filename
   * @returns Store identifier string
   */
  protected override extractStoreIdentifierFromFilename(filename: string): string {
    // Decode URL encoding first (filenames may be URL-encoded)
    const decodedFilename = decodeURIComponent(filename)

    // Match pattern like: PJ50-1, PJ7B-1, PJ8A-1 before the date (8 digits)
    // Store IDs are: PJ + alphanumeric + optional dash + digit
    const match = decodedFilename.match(/(PJ[\dA-Z]+-\d+)-\d{8}-\d{6}\.csv$/i)
    if (match) {
      return match[1]
    }

    // Try simpler pattern: PJ followed by alphanumeric characters
    const simpleMatch = decodedFilename.match(/(PJ[\dA-Z]+)-\d+-\d{8}/i)
    if (simpleMatch) {
      // Get the full store ID including the trailing number
      const afterPJ = decodedFilename.substring(decodedFilename.indexOf(simpleMatch[1]))
      const fullMatch = afterPJ.match(/^(PJ[\dA-Z]+-\d+)/i)
      if (fullMatch) {
        return fullMatch[1]
      }
      return simpleMatch[1]
    }

    // Last resort: use base class method
    return super.extractStoreIdentifierFromFilename(decodedFilename)
  }

  /**
   * Extract store metadata from KTC filename.
   *
   * KTC filenames follow the pattern:
   * TRGOVINA-ADDRESS-STORE_ID-DATE-TIME.csv
   *
   * Example: TRGOVINA-PAKRACKA ULICA 1 BJELOVAR-PJ50-1-20260105-071001.csv
   *
   * This method extracts:
   * - Store name: "KTC BJELOVAR" (from city)
   * - Address: "Pakracka Ulica 1" (excluding city)
   * - City: "Bjelovar" (last word of address)
   *
   * @param file - The discovered file
   * @returns Store metadata, or null if not extractable
   */
  extractStoreMetadata(file: DiscoveredFile): StoreMetadata | null {
    const decoded = decodeURIComponent(file.filename)

    // Extract address between TRGOVINA- and -PJ
    const match = decoded.match(/^TRGOVINA-(.+?)-(PJ[\dA-Z]+-\d+)-/i)
    if (!match) return super.extractStoreMetadata(file)

    const addressFull = match[1] // "PAKRACKA ULICA 1 BJELOVAR"
    // match[2] contains store code (e.g., "PJ50-1"), but we use city from address for naming

    // Last word is typically the city
    const words = addressFull.split(' ')
    const city = words.pop()
    const address = words.join(' ')

    return {
      name: `KTC ${this.titleCase(city || addressFull)}`,
      address: this.titleCase(address),
      city: city ? this.titleCase(city) : undefined,
    }
  }

  /**
   * Convert a string to title case.
   * Capitalizes the first letter of each word.
   *
   * @param str - The string to convert
   * @returns Title-cased string
   */
  private titleCase(str: string): string {
    return str
      .toLowerCase()
      .split(' ')
      .map((w) => w.charAt(0).toUpperCase() + w.slice(1))
      .join(' ')
  }
}

/**
 * Create a KTC adapter instance.
 */
export function createKtcAdapter(): KtcAdapter {
  return new KtcAdapter()
}

File: services/price-service/internal/adapters/chains/ktc.go
package chains

import (
	"fmt"
	"io"
	"net/url"
	"regexp"
	"strings"
	"time"

	"github.com/rs/zerolog/log"
	"github.com/kosarica/price-service/internal/adapters/base"
	"github.com/kosarica/price-service/internal/adapters/config"
	"github.com/kosarica/price-service/internal/parsers/csv"
	"github.com/kosarica/price-service/internal/types"
)

// ktcColumnMapping is the primary column mapping for KTC CSV files
var ktcColumnMapping = csv.CsvColumnMapping{
	ExternalID:     types.StringPtr("Šifra proizvoda"),
	Name:           "Naziv proizvoda",
	Category:       types.StringPtr("Kategorija"),
	Brand:          types.StringPtr("Marka proizvoda"),
	Unit:           types.StringPtr("Jedinica mjere"),
	UnitQuantity:   types.StringPtr("Neto količina"),
	Price:          "Maloprodajna cijena",
	DiscountPrice:  types.StringPtr("MPC za vrijeme posebnog oblika prodaje"),
	Barcodes:       types.StringPtr("Barkod"),
	UnitPrice:      types.StringPtr("Cijena za jedinicu mjere"),
	LowestPrice30d: types.StringPtr("Najniža cijena u posljednjih 30 dana"),
}

// ktcColumnMappingAlt is the alternative column mapping for KTC CSV files
var ktcColumnMappingAlt = csv.CsvColumnMapping{
	ExternalID:     types.StringPtr("Sifra proizvoda"),
	Name:           "Naziv proizvoda",
	Category:       types.StringPtr("Kategorija"),
	Brand:          types.StringPtr("Marka proizvoda"),
	Unit:           types.StringPtr("Jedinica mjere"),
	UnitQuantity:   types.StringPtr("Neto kolicina"),
	Price:          "Maloprodajna cijena",
	DiscountPrice:  types.StringPtr("MPC za vrijeme posebnog oblika prodaje"),
	Barcodes:       types.StringPtr("Barkod"),
	UnitPrice:      types.StringPtr("Cijena za jedinicu mjere"),
	LowestPrice30d: types.StringPtr("Najniza cijena u posljednjih 30 dana"),
}

// KtcAdapter is the chain adapter for KTC retail chain
type KtcAdapter struct {
	*base.BaseCsvAdapter
	discoveryDate string // Date filter for discovery (YYYY-MM-DD format)
}

// NewKtcAdapter creates a new KTC adapter
func NewKtcAdapter() (*KtcAdapter, error) {
	chainConfig := config.ChainConfigs[config.ChainKtc]

	adapterConfig := base.CsvAdapterConfig{
		BaseAdapterConfig: base.BaseAdapterConfig{
			Slug:           string(config.ChainKtc),
			Name:           chainConfig.Name,
			SupportedTypes: []types.FileType{types.FileTypeCSV},
			ChainConfig:    chainConfig,
			FilenamePrefixPatterns: []string{
				`(?i)^TRGOVINA[_-]?`,
				`(?i)^KTC[_-]?`,
				`(?i)^cjenik[_-]?`,
			},
		},
		ColumnMapping:            ktcColumnMapping,
		AlternativeColumnMapping: &ktcColumnMappingAlt,
	}

	baseAdapter, err := base.NewBaseCsvAdapter(adapterConfig)
	if err != nil {
		return nil, fmt.Errorf("failed to create base CSV adapter: %w", err)
	}

	return &KtcAdapter{
		BaseCsvAdapter: baseAdapter,
	}, nil
}

// SetDiscoveryDate sets the date to use for discovery filtering
func (a *KtcAdapter) SetDiscoveryDate(date string) {
	a.discoveryDate = date
}

// extractDateFromFilename extracts date from KTC filename
// Pattern: YYYYMMDD-HHMMSS at end of filename
func (a *KtcAdapter) extractDateFromFilename(filename string) string {
	// Match 8-digit date pattern (YYYYMMDD) before 6-digit time
	match := regexp.MustCompile(`(\d{4})(\d{2})(\d{2})-\d{6}\.csv$`).FindStringSubmatch(filename)
	if len(match) >= 4 {
		return fmt.Sprintf("%s-%s-%s", match[1], match[2], match[3])
	}
	return ""
}

// Discover discovers available KTC price files
// KTC uses a two-level portal: main page lists stores, each store page lists CSV files
func (a *KtcAdapter) Discover(targetDate string) ([]types.DiscoveredFile, error) {
	discoveredFiles := make([]types.DiscoveredFile, 0)
	seenURLs := make(map[string]bool)

	log.Debug().Str("chain", a.Name()).Str("portal", a.BaseURL()).Msg("Fetching portal")

	// Fetch main page to get list of stores
	resp, err := a.HTTPClient().Get(a.BaseURL())
	if err != nil {
		log.Error().Err(err).Msg("Failed to fetch KTC portal")
		return nil, fmt.Errorf("failed to fetch KTC portal: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		log.Error().Int("status_code", resp.StatusCode).Msg("KTC portal returned unexpected status")
		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
	}

	bodyBytes, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response body: %w", err)
	}

	html := string(bodyBytes)

	// Extract store names from links like: ?poslovnica=RC%20BJELOVAR%20PJ-50
	storePattern := regexp.MustCompile(`poslovnica=([^"&]+)`)
	matches := storePattern.FindAllStringSubmatch(html, -1)

	stores := make([]string, 0)
	seenStores := make(map[string]bool)
	for _, match := range matches {
		if len(match) >= 2 {
			storeName, _ := url.QueryUnescape(match[1])
			if !seenStores[storeName] {
				seenStores[storeName] = true
				stores = append(stores, storeName)
			}
		}
	}

	log.Debug().Int("store_count", len(stores)).Msg("Found stores on KTC portal")

	// For each store, fetch store page and extract CSV links
	for _, storeName := range stores {
		storeURL := fmt.Sprintf("%s?poslovnica=%s", a.BaseURL(), url.QueryEscape(storeName))

		storeResp, err := a.HTTPClient().Get(storeURL)
		if err != nil {
			log.Warn().Str("store", storeName).Err(err).Msg("Failed to fetch store page")
			continue
		}
		defer storeResp.Body.Close()

		if storeResp.StatusCode != 200 {
			log.Warn().Str("store", storeName).Int("status_code", storeResp.StatusCode).Msg("Store page returned unexpected status")
			continue
		}

		storeBodyBytes, err := io.ReadAll(storeResp.Body)
		if err != nil {
			continue
		}

		storeHtml := string(storeBodyBytes)

		// Extract CSV links like: /ktcftp/Cjenici/STORE_NAME/FILENAME.csv
		csvPattern := regexp.MustCompile(`href="([^"]*\.csv)"`)
		csvMatches := csvPattern.FindAllStringSubmatch(storeHtml, -1)

		for _, csvMatch := range csvMatches {
			if len(csvMatch) < 2 {
				continue
			}

			href := csvMatch[1]
			fileURL := href
			if !strings.HasPrefix(href, "http") {
				fileURL = a.BaseURL() + "/" + strings.TrimPrefix(href, "/")
			}

			// Skip duplicates
			if seenURLs[fileURL] {
				continue
			}
			seenURLs[fileURL] = true

			filename := a.extractFilenameFromURL(fileURL)
			fileDate := a.extractDateFromFilename(filename)

			// Filter by date if discoveryDate is set
			if a.discoveryDate != "" && fileDate != "" && fileDate != a.discoveryDate {
				continue
			}

			var lastModified *time.Time
			if fileDate != "" {
				if t, err := time.Parse("2006-01-02", fileDate); err == nil {
					lastModified = &t
				}
			}

			discoveredFiles = append(discoveredFiles, types.DiscoveredFile{
				URL:          fileURL,
				Filename:     filename,
				Type:         types.FileTypeCSV,
				Size:         nil,
				LastModified: lastModified,
				Metadata: map[string]string{
					"source":       "ktc_portal",
					"discoveredAt": time.Now().Format(time.RFC3339),
					"storeName":    storeName,
					"portalDate":   fileDate,
				},
			})
		}
	}

	return discoveredFiles, nil
}

// extractFilenameFromURL extracts filename from URL
func (a *KtcAdapter) extractFilenameFromURL(fileURL string) string {
	parsed, err := url.Parse(fileURL)
	if err != nil {
		return fileURL
	}

	path := parsed.Path
	parts := strings.Split(path, "/")
	if len(parts) > 0 {
		return parts[len(parts)-1]
	}

	return fileURL
}

// ExtractStoreIdentifierFromFilename extracts store identifier from KTC filename
// Pattern: TRGOVINA-ADDRESS-STORE_ID-DATE-TIME.csv
// Store IDs like: PJ50-1, PJ7B-1, PJ8A-1
func (a *KtcAdapter) ExtractStoreIdentifierFromFilename(filename string) string {
	// Try to match PJ followed by alphanumeric + dash + digit before date
	match := regexp.MustCompile(`(PJ[\dA-Z]+-\d+)-\d{8}-\d{6}\.csv$`).FindStringSubmatch(filename)
	if len(match) >= 2 {
		return match[1]
	}

	// Try simpler pattern: PJ followed by alphanumeric
	simpleMatch := regexp.MustCompile(`(PJ[\dA-Z]+)-\d+-\d{8}`).FindStringSubmatch(filename)
	if len(simpleMatch) >= 2 {
		return simpleMatch[1]
	}

	// Fallback to base class method
	dummyFile := types.DiscoveredFile{Filename: filename}
	if id := a.ExtractStoreIdentifier(dummyFile); id != nil {
		return id.Value
	}
	return ""
}

// ExtractStoreMetadata extracts store metadata from KTC filename
// Pattern: TRGOVINA-ADDRESS-STORE_ID-DATE-TIME.csv
func (a *KtcAdapter) ExtractStoreMetadata(file types.DiscoveredFile) *types.StoreMetadata {
	baseName := regexp.MustCompile(`\.(csv|CSV)$`).ReplaceAllString(file.Filename, "")

	// Extract address between TRGOVINA- and -PJ
	match := regexp.MustCompile(`^TRGOVINA-(.+?)-(PJ[\dA-Z]+-\d+)-`).FindStringSubmatch(baseName)
	if len(match) == 0 {
		storeID := a.ExtractStoreIdentifierFromFilename(file.Filename)
		return &types.StoreMetadata{
			Name: fmt.Sprintf("KTC %s", storeID),
		}
	}

	addressFull := match[1]

	// Last word is typically the city
	words := strings.Fields(addressFull)
	city := ""
	address := addressFull
	if len(words) > 1 {
		city = words[len(words)-1]
		address = strings.Join(words[:len(words)-1], " ")
	}

	return &types.StoreMetadata{
		Name:   fmt.Sprintf("KTC %s", titleCase(city)),
		Address: titleCase(address),
		City:   titleCase(city),
	}
}
